### ISSUE TITLE
Streaming Vectorizers / Chunking

### ISSUE THREAD

**Issue by kyleabeauchamp** [rmcgibbo/msmbuilder3#2]
So I know that we now have an interface for `updating` a model with new data, but have we figured out how we're going to handle actually chunking and iterating over the data?

For example, with atompairs, chunking at the single trajectory level will still be a real memory hog for medium sized proteins and trajectories.  

I guess the classes we want is something like `TrajectoryChunker` and `ProjectChunker` that provide a way to yield a bit of data at a time?

**Comment by kyleabeauchamp** [comment]
I guess the inputs to TrajectorChunker would be a single trajectory and the number of frames to keep in memory at once--which could be estimated by a simple rule of thumb (~no more than 20% RAM etc).

**Comment by rmcgibbo** [comment]
Did you see the yield_transform methods in the apps?

**Comment by kyleabeauchamp** [comment]
That handles trajectory loading, but not chunking, correct?

**Comment by rmcgibbo** [comment]
I think it can handle either, although we haven't implemented any chunking yet. But, for example, TICAApp implements `yield_transform`, so after the model is fit it can yield its data chunk-at-a-time. Whether those chunks amount to a single complete trajectory or not is not so essential.

**Comment by kyleabeauchamp** [comment]
I guess my only other question is whether this could somehow be done at the Pipeline level, rather than at the app level.  I guess that would be harder, as it would require propogating a chain of yields / fits throughout the pipeline.

**Comment by rmcgibbo** [comment]
Its possible, and that would be preferable. What I haven't figured out yet is how to label the classes so that you can manage the chunking/parallelism. Like, KCenter requires having the full dataset at once while tICA can be fit incrementally, etc. Managing all of this within pipeline is not trivial.

**Comment by rmcgibbo** [comment]
Do you like using the carat character to indicate the construction of a pipelined estimator? I will need to redo a lot of the command-line infastructure, but I think that will provide the impetus for this.

Basically, if you see the pipe character, the code builds a Pipeline object internally. And then we'll just have to solve this problem -- building a full-featured Pipeline.

**Comment by kyleabeauchamp** [comment]
I'm not sure whether I like ^ or ++ more, actually.  Your call.

**Comment by kyleabeauchamp** [comment]
PS: don't over-generalize things for my sake, I'm mostly playing devil's advocate here...

**Comment by rmcgibbo** [comment]
I'm not sure whether I like either. We could always do a short config file for pipelined estimators. That might be simpler. If the syntax that we make here is really designed to produce commands that are like over 100 characters, then maybe we're doing it wrong.

**Comment by kyleabeauchamp** [comment]
So my thoughts are:
1.  Make a library that is as easy to use as possible
2.  Find some what to make the library conveniently accessible from the command line.

Another possibility is to make a gui or interactive command line tool that asks the user a list of questions, then spits out a python script for them to run.  Much like your openmm web app.

**Comment by rmcgibbo** [comment]
I agree with both (1) and (2), but I think we need a command line interface as well. Ideally the command line interface is both functional and thin -- that encourages us to put most of the effort on the API.

What if there are two ways to use the command line:
- Specify a single method on the command line, like `msmb tICA -i inputvectors.h5 --n_components 10 -o outputvectors.h5`
- Specify more complex chained estimators using a short config file. like

```
$ cat ticamsm.json
[ 
  {'vector': {'option1' : 'a', 'option2': 'b'}},
  {'tICA': {'n_components': 2}},
  {'kcenters': {'n_clusters': 100}},
  {'msm': {'trim': true, 'symmetrize': 'mle'}}
]
```

If you want the simple interface, you run each step one command at a time with input files and output files. This exposes the full functionality of the package. Every command is tied to an individual estimator -- the command wrapping just adds a -i and -o for input and output paths.

**Comment by kyleabeauchamp** [comment]
That would work. 

My only complaint with a json file is that we're essentially just "leaving out the syntax"--we're using the CLI or python interface, but dropping the syntax.  To me, leaving out the syntax actually makes things seem more cryptic.  

I think at the end of the day, _any_ of these approaches will work.  We're trying to optimally wrap a very complex pipeline with a simple interface--there's no perfect solution, as we're balancing "power" and "ease".

**Comment by rmcgibbo** [comment]
What do you mean when you say "dropping the syntax"?

**Comment by kyleabeauchamp** [comment]
Well, in the CLI we would have a help menu and flags "--method tICA", or in Python we would have `model = Pipeline([Vector(), TICA(), kcenters()])`.

In the json, we don't really have any "context".

**Comment by rmcgibbo** [comment]
This is a general problem with config files, no? I think it's something that we can solve by having both the 1-step CLI and the multistep config file, by keeping them locked in a tight 1-1 mapping. So

```
$ cat vector.json
[{'vector': {'option1': 1}},
 {'kcenters': {'n_clusters': 10}}]
$ msmb -i trajectories/ -c vector.json -o output.h5
```

and

```
$ msmb vector -i trajectories --option1 1 -o temp.h5
$ msmb kcenters -i temp.h5 --n_clusters 10 -o output.h5
```

do _exactly_ the same thing. This way, if you're writing the config file, you can always go back and check using the CLI what the available options are.

**Comment by kyleabeauchamp** [comment]
That sounds pretty reasonable for users and for us to develop.

**Comment by rmcgibbo** [comment]
One thing that doesn't fit into this pattern is how to save the output data vs. the trained model itself. Like, there's serializing the kcenters object vs. serializing the labels for a dataset.

**Comment by schwancr** [comment]
Robert and I had an extended conversation regarding all of this and came up with a few good ideas. 

First, here are a list of advanced objects that are useful for the construction process:
1. Pipeline - (you guys already had this one). It has methods called (fit, transform, fit_transform, and transform_iter) depending on the end model (for instance it will only have a fit, if the last model is only an Estimator. The transform_iter is used to iterate through a dataset. Each model would implement this as well, allowing for a memory-efficient way to access the entire dataset, if necessary.
2. Fan (we had a better name, but I can't remember it) - This is like GridSearchCV in sklearn, which corresponds to a single model class, but a set of parameters. For example, there could be:
   `Fan(MSM, lag_times=[5, 10, 15, ..])`
   This would allow the user to calculate implied timescales, or just build many parallel MSMs. The result after fitting the fan, would be to use the `iter_models` method that returns an iterator over the fit models. The other methods would be some plural version of the underlying model (fit_models, transform_all; I don't think we decided on good names for these methods yet)
3. Workflow - This is a set of estimators, transformers, pipelines, and fans with a particular dependency graph. This would be the real workhorse for a user interacting via a python script. The idea here is that we often want to search over many parameters sets, but without redoing work. This could look something like:
   
   ```
   work = Workflow([Fan(vectorizers), 
                        Fan(tica, lag_times=[25, 50, 100]), 
                        Fan(kcenters, num_states=[5000, 10000]),
                        MSM('mle')])
   ```
   
   This would first transform a dataset according to each vectorizer in "vectorizers" then for each transformation, fit the tICA object with three different lag times, and so on. Basically each layer is applied to all leaves that were created from the previous step. 
   
   This interface allows for the most efficient use of your data. Under the hood, however, it's does not need to be a simple for-loop. We can add MPI support, which would allow you to submit one job to a cluster that produces many MSMs. 
   
   The workflow object would have a simplified set of methods (maybe just something like `run()`) since its functionality is quite diverse. There are additional details to be hammered out here (for instance storing an entire workflow in memory is probably impossible), but it would be very powerful if we can get it to work.

Due to the potential complexity (due to their inherent flexibility) of these objects, we don't want to provide a command line interface to them. The command line would still be there, but would be limited to single models/estimators (tICA, vectorize, MSM, etc.). In this way, if someone does not want to write python, they can still write a workflow in a bash script and save the intermediate data. 

What do you think about these ideas?

**Comment by kyleabeauchamp** [comment]
This sounds good to me, but also a pretty substantial engineering challenge.

**Comment by schwancr** [comment]
One problem that @rmcgibbo and I were discussing was how to construct a branch (called a fan above) if there are several classes.

One way would look like this:

```
    b = Branch()
    b.add(DihedralVectorizer, {'angles' : angle_list})
    b.add(DistanceVectorizer, {'contacts' : contact_list})
```

The other way would be to be able to provide a list of classes, as well as a list of `param_grid`s. In sklearns GridSearchCV, the param_grid is actually a list of dicts already, and so making that kwarg a list of lists could get confusing and unsightly.

**Comment by kyleabeauchamp** [comment]
This looks fine.

PS: I wonder if it makes sense for us to work out a triage chart for our wishlist--a google doc spreadsheet or something.  I say this because we have a lot of "wants" for MSMB3, and it might be useful for us to attach a cost to each item.  We could also assign tasks etc.  We "kinda" already have this with GitHub, but it might be good to have a "grid" style view.

**Comment by rmcgibbo** [comment]
We could do that. And/or people can just step up and get stuff done. In order to avoid two people doing the same thing, you can create a WIP PR on the first commit.
